derive/src/utils.rs

```diff
@@ -45,10 +45,10 @@ pub fn const_eval_check_variant_indexes(
 	let mut recurse_indices = vec![];
 	for (ident, index) in recurse_variant_indices {
 		let ident_str = ident.to_string();
-		// We convert to u8 same as in the generated code.
+		// We convert to usize same as in the generated code.
 		recurse_indices.push(quote_spanned! { ident.span() =>
 			(
-				(#index) as ::core::primitive::u8,
+				(#index) as ::core::primitive::usize,
```

Comment:
I think u8 is more accurate, in the implementation of decode and encode, the type is cast to u8, if user writes:
```rust
enum Foo {
  A = 256,
  B = 0,
}
```
then we want to give an error for conflicts.

The code in scale codec could be refactored to put the cast as part of the expression for variant indices but I think we are fine.

Reply:
We have an extra check for this [here](https://github.com/paritytech/parity-scale-codec/blob/49995c8ceeba6785c227e2d5b611e8a49dd5a818/derive/src/utils.rs#L523)

Reply:
This doesn't take the discriminant into account no?

Reply:
That is a good point. I will add a check there, still think this is better than having it fail somewhere with a weird error.

Reply:
https://github.com/paritytech/parity-scale-codec/blob/49995c8ceeba6785c227e2d5b611e8a49dd5a818/derive/src/utils.rs#L493 ha we have this :D

Reply:
@gui1117 please have another look